### PR TITLE
JDK-8284178: os::commit_memory() should assert the given range

### PR DESCRIPTION
**Pull request by tstuefe**
Trivial fix to assert that ranges given to os::commit/uncommit/release_memory are not empty.

I recently hunted an error where the memory reservation failed, was not handled, and the null pointer was accidentally passed to os::commit_memory(). The resulting `mmap()` failed and this looked like a commit error, when it really was a reservation error. An assert would have saved me time.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8284178](https://bugs.openjdk.java.net/browse/JDK-8284178): os::commit_memory() should assert the given range

### Reviewers
 * [Aleksey Shipilev](https://openjdk.java.net/census#shade) (@shipilev - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8075/head:pull/8075` \
`$ git checkout pull/8075`

Update a local copy of the PR: \
`$ git checkout pull/8075` \
`$ git pull https://git.openjdk.java.net/jdk pull/8075/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8075`

View PR using the GUI difftool: \
`$ git pr show -t 8075`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8075.diff">https://git.openjdk.java.net/jdk/pull/8075.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/runtime/os.cpp [hunk 1]
```diff
@@ -1675,7 +1675,13 @@ char* os::attempt_reserve_memory_at(char* addr, size_t bytes, bool executable) {
   return result;
 }
 
+static void assert_nonempty_range(const char* addr, size_t bytes) {
+  assert(addr != nullptr && bytes > 0, "invalid range (" PTR_FORMAT ", " SIZE_FORMAT ")",
```
Comment (shipilev) [inline src/hotspot/share/runtime/os.cpp:1679]: The message suggests it is *range*, i.e. `(begin, end)`. I wonder if we better print `p2i(addr) + bytes`?

**Review by shipilev — CHANGES_REQUESTED** [review]

File: src/hotspot/share/runtime/os.cpp (hunk 1 above)
Comment (tstuefe) [reply src/hotspot/share/runtime/os.cpp]: Okay, I changed the message

Comment (shipilev) [reply src/hotspot/share/runtime/os.cpp]: I think `p2i(addr + bytes)` is UB when `addr` is `nullptr`? `p2i(addr) + bytes` is safer?

File: src/hotspot/share/runtime/os.cpp [hunk 2]
```diff
@@ -1675,7 +1675,13 @@ char* os::attempt_reserve_memory_at(char* addr, size_t bytes, bool executable) {
   return result;
 }
 
+static void assert_nonempty_range(const char* addr, size_t bytes) {
+  assert(addr != nullptr && bytes > 0, "invalid range [" PTR_FORMAT ", " PTR_FORMAT ")",
+         p2i(addr), p2i(addr + bytes));
```
Comment (shipilev) [inline src/hotspot/share/runtime/os.cpp:1680]: I think `p2i(addr + bytes)` is UB when `addr` is `nullptr`? `p2i(addr) + bytes` is safer?

Comment (tstuefe) [reply src/hotspot/share/runtime/os.cpp]: You are right. I fixed it.

**Review by shipilev — APPROVED** [review]

**Comment by tstuefe** [comment]
@shipilev would you agree this is trivial?

**Comment by shipilev** [comment]
> @shipilev would you agree this is trivial?

Not trivial, but simple enough to go in with one review, I think.

**Comment by tstuefe** [comment]
> > @shipilev would you agree this is trivial?
> 
> Not trivial, but simple enough to go in with one review, I think.

Thanks Aleksey.

/integrate

**Pull request closed without merging**